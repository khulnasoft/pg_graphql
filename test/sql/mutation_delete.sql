begin;

    create table account(
        id serial primary key,
        email varchar(255) not null
    );

    create function _echo_email(account)
        returns text
        language sql
    as $$ select $1.email $$;

    create table blog(
        id serial primary key,
        owner_id integer not null references account(id) on delete cascade,
        name varchar(255) not null
    );

    insert into public.account(email)
    values
        ('aardvark@x.com'),
        ('bat@x.com'),
        ('cat@x.com'),
        ('dog@x.com'),
        ('elephant@x.com');

    insert into blog(owner_id, name)
    values
        (1, 'A: Blog 1'),
        (1, 'A: Blog 2'),
        (2, 'A: Blog 3'),
        (2, 'B: Blog 3');
    comment on table blog is e'@graphql({"totalCount": {"enabled": true}})';

    savepoint a;

    -- Check atMost clause stops deletes
    select graphql.resolve($$
    mutation {
      deleteFromAccountCollection(
        filter: {
          email: {eq: "bat@x.com"}
        }
        atMost: 0
      ) {
        affectedCount
        records {
          id
          email
          echoEmail
          blogCollection {
            totalCount
            edges {
              node {
                id
              }
            }
          }
        }
      }
    }
    $$);

    rollback to savepoint a;

    -- Check delete works and allows nested response
    select jsonb_pretty(
        graphql.resolve($$
            mutation {
              deleteFromAccountCollection(
                filter: {
                  email: {eq: "bat@x.com"}
                }
                atMost: 1
              ) {
                affectedCount
                records {
                  id
                  email
                  echoEmail
                  blogCollection {
                    totalCount
                    edges {
                      node {
                        id
                      }
                    }
                  }
                }
              }
            }
        $$)
    );

    rollback to savepoint a;

    -- Check `atMost` clause can be omitted b/c of default
    select jsonb_pretty(
        graphql.resolve($$
            mutation {
              deleteFromAccountCollection(
                filter: {
                  email: {eq: "bat@x.com"}
                }
              ) {
                records { id }
              }
            }
        $$)
    );

    rollback to savepoint a;

    -- Check no matches returns empty array vs null + allows top xyz alias
    select jsonb_pretty(
        graphql.resolve($$
            mutation {
              xyz: deleteFromAccountCollection(
                filter: {
                  email: {eq: "no@match.com"}
                }
                atMost: 1
              ) {
                records { id }
              }
            }
        $$)
    );

    rollback to savepoint a;

    -- Check no filter deletes all records
    select jsonb_pretty(
        graphql.resolve($$
            mutation {
              deleteFromAccountCollection(
                atMost: 8
              ) {
                records { id }
              }
            }
        $$)
    );

rollback;
