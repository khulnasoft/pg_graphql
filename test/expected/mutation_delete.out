begin;
    create table account(
        id serial primary key,
        email varchar(255) not null
    );
    create function _echo_email(account)
        returns text
        language sql
    as $$ select $1.email $$;
    create table blog(
        id serial primary key,
        owner_id integer not null references account(id) on delete cascade,
        name varchar(255) not null
    );
    insert into public.account(email)
    values
        ('aardvark@x.com'),
        ('bat@x.com'),
        ('cat@x.com'),
        ('dog@x.com'),
        ('elephant@x.com');
    insert into blog(owner_id, name)
    values
        (1, 'A: Blog 1'),
        (1, 'A: Blog 2'),
        (2, 'A: Blog 3'),
        (2, 'B: Blog 3');
    comment on table blog is e'@graphql({"totalCount": {"enabled": true}})';
    savepoint a;
    -- Check atMost clause stops deletes
    select graphql.resolve($$
    mutation {
      deleteFromAccountCollection(
        filter: {
          email: {eq: "bat@x.com"}
        }
        atMost: 0
      ) {
        affectedCount
        records {
          id
          email
          echoEmail
          blogCollection {
            totalCount
            edges {
              node {
                id
              }
            }
          }
        }
      }
    }
    $$);
                                  resolve                                   
----------------------------------------------------------------------------
 {"data": null, "errors": [{"message": "delete impacts too many records"}]}
(1 row)

    rollback to savepoint a;
    -- Check delete works and allows nested response
    select jsonb_pretty(
        graphql.resolve($$
            mutation {
              deleteFromAccountCollection(
                filter: {
                  email: {eq: "bat@x.com"}
                }
                atMost: 1
              ) {
                affectedCount
                records {
                  id
                  email
                  echoEmail
                  blogCollection {
                    totalCount
                    edges {
                      node {
                        id
                      }
                    }
                  }
                }
              }
            }
        $$)
    );
                 jsonb_pretty                  
-----------------------------------------------
 {                                            +
     "data": {                                +
         "deleteFromAccountCollection": {     +
             "records": [                     +
                 {                            +
                     "id": 2,                 +
                     "email": "bat@x.com",    +
                     "echoEmail": "bat@x.com",+
                     "blogCollection": {      +
                         "edges": [           +
                             {                +
                                 "node": {    +
                                     "id": 3  +
                                 }            +
                             },               +
                             {                +
                                 "node": {    +
                                     "id": 4  +
                                 }            +
                             }                +
                         ],                   +
                         "totalCount": 2      +
                     }                        +
                 }                            +
             ],                               +
             "affectedCount": 1               +
         }                                    +
     }                                        +
 }
(1 row)

    rollback to savepoint a;
    -- Check `atMost` clause can be omitted b/c of default
    select jsonb_pretty(
        graphql.resolve($$
            mutation {
              deleteFromAccountCollection(
                filter: {
                  email: {eq: "bat@x.com"}
                }
              ) {
                records { id }
              }
            }
        $$)
    );
               jsonb_pretty               
------------------------------------------
 {                                       +
     "data": {                           +
         "deleteFromAccountCollection": {+
             "records": [                +
                 {                       +
                     "id": 2             +
                 }                       +
             ]                           +
         }                               +
     }                                   +
 }
(1 row)

    rollback to savepoint a;
    -- Check no matches returns empty array vs null + allows top xyz alias
    select jsonb_pretty(
        graphql.resolve($$
            mutation {
              xyz: deleteFromAccountCollection(
                filter: {
                  email: {eq: "no@match.com"}
                }
                atMost: 1
              ) {
                records { id }
              }
            }
        $$)
    );
       jsonb_pretty       
--------------------------
 {                       +
     "data": {           +
         "xyz": {        +
             "records": [+
             ]           +
         }               +
     }                   +
 }
(1 row)

    rollback to savepoint a;
    -- Check no filter deletes all records
    select jsonb_pretty(
        graphql.resolve($$
            mutation {
              deleteFromAccountCollection(
                atMost: 8
              ) {
                records { id }
              }
            }
        $$)
    );
               jsonb_pretty               
------------------------------------------
 {                                       +
     "data": {                           +
         "deleteFromAccountCollection": {+
             "records": [                +
                 {                       +
                     "id": 1             +
                 },                      +
                 {                       +
                     "id": 2             +
                 },                      +
                 {                       +
                     "id": 3             +
                 },                      +
                 {                       +
                     "id": 4             +
                 },                      +
                 {                       +
                     "id": 5             +
                 }                       +
             ]                           +
         }                               +
     }                                   +
 }
(1 row)

rollback;
